### PR TITLE
Protect versus unbounded recursion

### PR DESCRIPTION
**Pull request by svanoort**
Solves #76 albeit in a rather crude way. 

I'd like to do this ultimately in a way that provides better and deeper diagnostic info (i.e. trying to resolve library names and a deeper stack trace within the CPS code), but this is the simplest thing that will resolve it here and now and help point to where to fix the code issue.

Snapshot deployed as: 1.21-20171018.220523-1

# Screenshot of the result when workflow-cps is run with this plugin version in an endlessly recursive situation
screen shot 2017-10-09 at 3 08 42 pm

### CODE REVIEW

File: lib/src/main/java/com/cloudbees/groovy/cps/impl/CallEnv.java [hunk 1]
```diff
@@ -57,6 +60,11 @@ public CallEnv(Env caller, Continuation returnAddress, SourceLocation loc, int l
         } else {
             types = Maps.newHashMapWithExpectedSize(localsCount);
         }
+        depth = (caller instanceof  DepthTrackingEnv) ? depth = ((DepthTrackingEnv) caller).getDepth() + 1 : 1;
+
+        if (depth > DepthTrackingEnv.MAX_LEGAL_DEPTH) {
+            throw new StackOverflowError("Excessively nested blocks/function calls exceed limit of"+DepthTrackingEnv.MAX_LEGAL_DEPTH+":"+depth);
```
Comment (alexanderrtaylor) [inline lib/src/main/java/com/cloudbees/groovy/cps/impl/CallEnv.java]: need a space after "of". But do you think there is a way in here to point to the user's code?

Comment (svanoort) [reply lib/src/main/java/com/cloudbees/groovy/cps/impl/CallEnv.java]: Yes, I have a commit locally that fixes the space issue (pushing it publicly now) and I'm working on how to point the user's code at it

**Comment by svanoort** [comment]
@reviewbybees

File: lib/src/main/java/com/cloudbees/groovy/cps/CpsTransformer.java [hunk 2]
```diff
@@ -239,8 +240,9 @@ CpsFunction ___cps___N() {
                 new StaticMethodCallExpression(m.getDeclaringClass(), cpsName, new TupleExpression()));
 //                new ConstructorCallExpression(FUNCTION_TYPE, new TupleExpression(params, body)));
 
-        List<Expression> paramExpressions = new ArrayList<>();
-        for (Parameter p : m.getParameters()) {
+        Parameter[] pms = m.getParameters();
+        List<Expression> paramExpressions = new ArrayList<>(pms.length);
+        for (Parameter p : pms) {
```
Comment (jglick) [inline lib/src/main/java/com/cloudbees/groovy/cps/CpsTransformer.java]: It looks like all the changes to this file are some unrelated optimization. If so, would better be filed as a separate PR.

File: lib/src/main/java/com/cloudbees/groovy/cps/DepthTrackingEnv.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package com.cloudbees.groovy.cps;
+
+/**
+ * @author Sam Van Oort
+ */
+public interface DepthTrackingEnv extends Env {
+
+    /** Limit on how deeply environments can recurse */
+    int MAX_LEGAL_DEPTH = 3000;
```
Comment (jglick) [inline lib/src/main/java/com/cloudbees/groovy/cps/DepthTrackingEnv.java]: BTW how does this compare to default stack size for a typical JVM and effective recursion depth in that case?

File: lib/src/main/java/com/cloudbees/groovy/cps/DepthTrackingEnv.java [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+    public int getDepth();
```
Comment (jglick) [inline lib/src/main/java/com/cloudbees/groovy/cps/DepthTrackingEnv.java]: Javadoc please—the return value here is by no means self-explanatory.

File: lib/src/main/java/com/cloudbees/groovy/cps/impl/CallEnv.java [hunk 5]
```diff
@@ -38,6 +39,8 @@
 
     private Invoker invoker;
 
+    int depth;
```
Comment (jglick) [inline lib/src/main/java/com/cloudbees/groovy/cps/impl/CallEnv.java]: After deserialization of a program running prior to the upgrade, this will be 0. Is that expected and safe?

File: lib/src/main/java/com/cloudbees/groovy/cps/impl/CallEnv.java [hunk 6]
```diff
@@ -57,6 +60,15 @@ public CallEnv(Env caller, Continuation returnAddress, SourceLocation loc, int l
         } else {
             types = Maps.newHashMapWithExpectedSize(localsCount);
         }
+        depth = (caller instanceof  DepthTrackingEnv) ? depth = ((DepthTrackingEnv) caller).getDepth() + 1 : 1;
```
Comment (jglick) [inline lib/src/main/java/com/cloudbees/groovy/cps/impl/CallEnv.java]: Using return value of assignment, confusingly and probably incorrectly.

File: lib/src/main/java/com/cloudbees/groovy/cps/impl/CallEnv.java [hunk 7]
```diff
@@ -57,6 +60,15 @@ public CallEnv(Env caller, Continuation returnAddress, SourceLocation loc, int l
...
+
+        if (depth > DepthTrackingEnv.MAX_LEGAL_DEPTH) {
+            if (loc != null) {
+                throw new StackOverflowError("Excessively nested closures/functions at "+loc.toString()+" - look for unbounded recursion - call depth: "+depth);
```
Comment (jglick) [inline lib/src/main/java/com/cloudbees/groovy/cps/impl/CallEnv.java]: BTW no need for `.toString()` here.

File: lib/src/main/java/com/cloudbees/groovy/cps/impl/CallEnv.java [hunk 8]
```diff
@@ -57,6 +60,15 @@ public CallEnv(Env caller, Continuation returnAddress, SourceLocation loc, int l
...
+            } else {
+                throw new StackOverflowError("Excessively nested blocks/function - look for unbounded recursion - call depth: "+depth);
```
Comment (jglick) [inline lib/src/main/java/com/cloudbees/groovy/cps/impl/CallEnv.java]: Blocks here but closures above? Intentional?

File: lib/src/main/java/com/cloudbees/groovy/cps/impl/ProxyEnv.java [hunk 9]
```diff
@@ -13,11 +14,14 @@
  *
  * @author Kohsuke Kawaguchi
  */
-public class ProxyEnv implements Env {
+public class ProxyEnv implements DepthTrackingEnv {
     protected final Env parent;
 
+    int depth = 0;
```
Comment (jglick) [inline lib/src/main/java/com/cloudbees/groovy/cps/impl/ProxyEnv.java]: Unused initializer—overwritten by constructor.

File: lib/src/test/groovy/com/cloudbees/groovy/cps/impl/FunctionCallBlockTest.groovy [hunk 10]
```diff
@@ -21,6 +22,28 @@ class FunctionCallBlockTest extends AbstractGroovyCpsTest {
...
+            assert evalCPSonly("""
+                def thing = null;
+                def getThing() { 
+                    return thing == null;
+                }
+                def stuff = getThing();
+            """, 25000) == "cheese";
+        } catch (StackOverflowError soe) {
+            println "PASSED: expected exception thrown for endlessly recursive function, see trace below";
+            soe.printStackTrace()
```
Comment (jglick) [inline lib/src/test/groovy/com/cloudbees/groovy/cps/impl/FunctionCallBlockTest.groovy]: Maybe assert that the detail message has some kind of useful diagnostic.

File: lib/src/test/groovy/com/cloudbees/groovy/cps/impl/FunctionCallBlockTest.groovy [hunk 11]
```diff
@@ -21,6 +22,28 @@ class FunctionCallBlockTest extends AbstractGroovyCpsTest {
...
+    void infiniteRecursion() {
+        try {
+            assert evalCPSonly("""
+                def thing = null;
+                def getThing() { 
+                    return thing == null;
+                }
+                def stuff = getThing();
+            """, 25000) == "cheese";
+        } catch (StackOverflowError soe) {
```
Comment (jglick) [inline lib/src/test/groovy/com/cloudbees/groovy/cps/impl/FunctionCallBlockTest.groovy]: You need to `fail` if the error is _not_ thrown.

File: lib/src/test/groovy/com/cloudbees/groovy/cps/impl/FunctionCallBlockTest.groovy [hunk 12]
```diff
@@ -21,6 +22,28 @@ class FunctionCallBlockTest extends AbstractGroovyCpsTest {
...
+
+    @Test
+    void infiniteRecursion() {
+        try {
+            assert evalCPSonly("""
+                def thing = null;
+                def getThing() { 
+                    return thing == null;
+                }
+                def stuff = getThing();
```
Comment (jglick) [inline lib/src/test/groovy/com/cloudbees/groovy/cps/impl/FunctionCallBlockTest.groovy]: Cf. the associated issue. It seems this _is_ a buggy program (reproducible using command-line Groovy—`def thing` does not declare a field), but it might be clearer for test purposes to use a self-evidently endless recursion, such as:

```groovy
def buggy() {
    buggy()
}
buggy()
```

File: lib/src/main/java/com/cloudbees/groovy/cps/impl/CallEnv.java (hunk 5 above)
Comment (svanoort) [reply lib/src/main/java/com/cloudbees/groovy/cps/impl/CallEnv.java]: Yes -- worst case, calls can run deeper than they're supposed to, but since the limit we set is arbitrary that's not a huge problem.  We just want to cover the unbounded cases.

File: lib/src/main/java/com/cloudbees/groovy/cps/DepthTrackingEnv.java (hunk 3 above)
Comment (svanoort) [reply lib/src/main/java/com/cloudbees/groovy/cps/DepthTrackingEnv.java]: Varies depending on environment and settings.  In common cases, could be as little as 1000, or as much as 8000. 

In practice really all we care about is that "sane" code will run and that we catch runaway recursion -- also serializing the program will generate a StackOverflowError at some point (probably south of 1000 runs).

File: lib/src/test/groovy/com/cloudbees/groovy/cps/impl/FunctionCallBlockTest.groovy (hunk 11 above)
Comment (svanoort) [reply lib/src/test/groovy/com/cloudbees/groovy/cps/impl/FunctionCallBlockTest.groovy]: Fails implicitly, the above never returns "cheese" ;)

**Comment by michaelneale** [comment]
nice sam!

<insert joke about solving the halting problem here>

**Comment by svanoort** [comment]
@michaelneale Thanks!  The solution is rounding, of course if you squint really hard or use an imprecise numeric system,  1024 is close enough to Infinity.

**Comment by michaelneale** [comment]
:bee:

**Comment by svanoort** [comment]
@abayer  @jglick Please may I request re-review of this fix from one of you -- made some changes because with the exception simply thrown pipeline wasn't cleaning up the executor like it ought to have -- verification test in https://github.com/jenkinsci/workflow-cps-plugin/pull/183 -- thanks!

File: lib/src/test/groovy/com/cloudbees/groovy/cps/impl/FunctionCallBlockTest.groovy (hunk 12 above)
Comment (michaelneale) [reply lib/src/test/groovy/com/cloudbees/groovy/cps/impl/FunctionCallBlockTest.groovy]: i don’t think def thing is needed at all to show this anyway.

**Comment by michaelneale** [comment]
Cool stuff Sam!

File: dgm-builder/pom.xml [hunk 13]
```diff
@@ -4,7 +4,7 @@
     <parent>
         <groupId>com.cloudbees</groupId>
         <artifactId>groovy-cps-parent</artifactId>
-        <version>1.21-SNAPSHOT</version>
+        <version>1.21-stop-recursion2-SNAPSHOT</version>
```
Comment (jglick) [inline dgm-builder/pom.xml]: Revert please. Not necessary for testing downstream anyway; just use the timestamp in the dependency.

File: lib/src/main/java/com/cloudbees/groovy/cps/DepthTrackingEnv.java [hunk 14]
```diff
@@ -5,8 +5,10 @@
  */
 public interface DepthTrackingEnv extends Env {
 
-    /** Limit on how deeply environments can recurse */
-    int MAX_LEGAL_DEPTH = 3000;
+    /** Limit on how deeply environments can recurse.
+     *  Capped somewhat low to try to limit the ability to run a program that will generate a StackOverflowError when serialized. */
+    int MAX_LEGAL_DEPTH = 1024;
```
Comment (jglick) [inline lib/src/main/java/com/cloudbees/groovy/cps/DepthTrackingEnv.java]: I am afraid this might kill legitimate programs which just happen to use, say, a recursive function to process a list or whatever. (Yes common Groovy idioms would typically make that unnecessary, but…)

File: lib/src/main/java/com/cloudbees/groovy/cps/impl/CpsClosureDef.java [hunk 15]
```diff
@@ -34,6 +35,19 @@ Next invoke(Env caller, SourceLocation loc, Object receiver, List<?> args, Conti
 
         // TODO: who handles 'it' ?
 
+        // 'e' will be a DepthTrackingEnv for all pipelines begun with this version of the library
+        // but if we upgrade groovy-cps and resume older pipelines, they won't be.
+        if (e instanceof DepthTrackingEnv && ((DepthTrackingEnv)e).getDepth() > DepthTrackingEnv.MAX_LEGAL_DEPTH) {
+            StackOverflowError err;
+            int depth = ((DepthTrackingEnv) e).getDepth();
+            if (loc != null) {
```
Comment (jglick) [inline lib/src/main/java/com/cloudbees/groovy/cps/impl/CpsClosureDef.java]: Maybe better to extract a utility function out of this (`@CheckForNull Next`).

File: lib/src/main/java/com/cloudbees/groovy/cps/impl/CpsClosureDef.java [hunk 16]
```diff
@@ -34,6 +35,19 @@ Next invoke(Env caller, SourceLocation loc, Object receiver, List<?> args, Conti
 
         // TODO: who handles 'it' ?
 
+        // 'e' will be a DepthTrackingEnv for all pipelines begun with this version of the library
+        // but if we upgrade groovy-cps and resume older pipelines, they won't be.
```
Comment (jglick) [inline lib/src/main/java/com/cloudbees/groovy/cps/impl/CpsClosureDef.java]: Is that actually true? `DepthTrackingEnv` is just a marker interface. Existing instances of classes which retroactively implement it will be `instanceof DepthTrackingEnv` after a restart.

I think what you meant to say is that `getDepth()` would be 0 in such cases, since the field would be null after deserialization.

Comment (svanoort) [reply lib/src/main/java/com/cloudbees/groovy/cps/impl/CpsClosureDef.java]: I'm honestly not 100% sure if we'll have a path that provides an Env that isn't a DepthTrackingEnv -- but I'm also cautious about things like this.  We can revise the comment if you like.

File: dgm-builder/pom.xml (hunk 13 above)
Comment (svanoort) [reply dgm-builder/pom.xml]: It'll be reverted before merge -- I'm using this for local testing and don't seem to have the capability to deploy a SNAPSHOT -- and this isn't part of the repository-permissions-updater.

File: lib/src/main/java/com/cloudbees/groovy/cps/DepthTrackingEnv.java (hunk 14 above)
Comment (michaelneale) [reply lib/src/main/java/com/cloudbees/groovy/cps/DepthTrackingEnv.java]: yes, but necessary

Comment (svanoort) [reply lib/src/main/java/com/cloudbees/groovy/cps/DepthTrackingEnv.java]: They'd run into a StackOverflowError eventually if they *weren't* NonCPS code though -- we're just setting an explicit threshold.  1024 is high enough that I feel comfortable saying "if you need more than this, you shouldn't be running it as part of CPS execution." 

There's always the escape valve of NonCPS annotations and ability to rewrite logic to use a non-recursive form.

File: lib/src/main/java/com/cloudbees/groovy/cps/impl/CpsClosureDef.java (hunk 16 above)
Comment (svanoort) [reply lib/src/main/java/com/cloudbees/groovy/cps/impl/CpsClosureDef.java]: You know what,  I don't know what I was thinking.  We don't need to worry about this at all, fixing it.

**Comment by jglick** [comment]
> don't seem to have the capability to deploy a SNAPSHOT

Mentioned out of band but FTR:

```bash
alias deploysnapshot='mvn clean install source:jar deploy:deploy -DaltDeploymentRepository=maven.jenkins-ci.org::default::https://repo.jenkins-ci.org/snapshots/'
```

File: lib/src/main/java/com/cloudbees/groovy/cps/impl/CpsFunction.java [hunk 17]
```diff
@@ -18,8 +19,18 @@ public CpsFunction(List<String> parameters, Block body) {
...
         assignArguments(args,e);
+
+        if (e.getDepth() > DepthTrackingEnv.MAX_LEGAL_DEPTH) {
+            StackOverflowError err;
+            if (loc != null) {
+                err = new StackOverflowError("Excessively nested closures/functions at "+loc+" - look for unbounded recursion - call depth: "+e.getDepth());
+            } else {
+                err = new StackOverflowError("Excessively nested closures/functions - look for unbounded recursion - call depth: "+e.getDepth());
+            }
+            return e.getExceptionHandler(StackOverflowError.class).receive(err);
```
Comment (jglick) [inline lib/src/main/java/com/cloudbees/groovy/cps/impl/CpsFunction.java]: Useful to extract a method for this stuff (DRY).

**Comment by svanoort** [comment]
I'm holding off merge-release until I can figure out wtf is going on with the Windows issue in https://github.com/jenkinsci/workflow-cps-plugin/pull/183#issuecomment-337769172

**Comment by svanoort** [comment]
Tests were just flaky (unrelated issues, addressed elsewhere) - so I'm going ahead to merge since this has been approved.
